Core 467/support argo rollouts

natyharary opened the pull request:
Support Argo Rollouts CRDs in the odigos.
This includes:
1. Support instrumentation and de-instrumentation of rollouts in the instrumentor
2. UI and CLI support for these new objects
3. Updated RBAC for the odigos controller and operator to be able to list and interact with rollouts
4. Support for both Argo Rollout startegies - green/blue and canary

Tests done:
🆗 Cluster without argo rollout controller works as usual - instrumentation, de-instrumentation with the demo app, restarts
🆗 Cluster with argo rollouts controller + namespace - deploying a blue/green argo rollout deployment with the python test app
🆗 For previous deployment, promote to have two replicasets: one blue, one green
🆗 Instrumentation of said argo rollout caused restart of both blue and green sets + instrumentation, without switching blue to green
🆗 De-istrumentation of said argo rollout caused restart of both blue and green sets + instrumentation, without switching blue to green
🆗 Work with the UI and CLI with argo rollouts proved satisfactory
🆗 Instrumentation + de-instrumentation of canary rollouts - start, end and mid-rollout
🆗 Disabling auto rollout and seeing that instrumentation and de-instrumentation of argo rollouts doesn't trigger a restart

emergency-ticket id: EMR-749

natyharary committed:
chore: update go.mods to include argo-rollouts v1.8.3

natyharary committed:
feat: add argo rollouts support in the instrumentor

natyharary committed:
feat: add argo rollout support in k8sutils + differentiating between replicaset owners

ReplicaSets can be managed from either deployment or argo rollouts, so we need determine which one is which

natyharary committed:
feat: add argo rollout support for the frontend

natyharary committed:
feat: update helm RBAC for handling rollouts

natyharary committed:
feat: add rollout support for cli

natyharary committed:
feat: add support for Argo Rollouts in metrics collection

- Included custom attribute for Argo Rollouts in the metrics UI configuration.
- Updated test data to reflect the addition of Argo Rollouts in the processors for logs and traces.

natyharary committed:
fix: canary rollouts mid-deployment will now be instrumented

natyharary committed:
fix: linting issues

natyharary committed:
fix: go-mod tidy

natyharary committed:
feat: enhance RBAC permissions for Argo Rollouts in operator

RonFed commented on api/k8sconsts/podworkload.go:
```diff
@@ -12,6 +12,7 @@ const (
 	WorkloadKindCronJob          WorkloadKind = "CronJob"
 	WorkloadKindJob              WorkloadKind = "Job"
 	WorkloadKindDeploymentConfig WorkloadKind = "DeploymentConfig"
+	WorkloadKindRollout          WorkloadKind = "Rollout"
```
nit: (for clarity)

```suggestion
	WorkloadKindArgoRollout          WorkloadKind = "ArgoRollout"
```

natyharary replied:
Agreed

RonFed commented on cli/cmd/resources/instrumentor.go:
```diff
@@ -256,6 +256,11 @@ func NewInstrumentorClusterRole(openshiftEnabled bool) *rbacv1.ClusterRole {
+				APIGroups: []string{"argoproj.io"},
+				Resources: []string{"rollouts"},
+				Verbs:     []string{"get", "list", "watch", "patch"},
+			},
```
I think this part of the code is deprecated, not sure if we want to add stuff here
cc @tamirdavid1

natyharary replied:
Tamir and I talked - we still want this code to be maintained as it is still partially used by the UI.

damemi replied:
+1 thanks @natyharary, Tamir talked to me too-- it's really just the operator that's still dependent on this so I appreciate everyone's patience while I find time to migrate that over to the new helm backed cli deploy method

RonFed commented on k8sutils/pkg/workload/ownerreference.go:
```diff
@@ -68,6 +71,16 @@ func GetWorkloadNameAndKind(ownerName, ownerKind string) (string, k8sconsts.Work
+// ReplicaSets can be created from either Deployment or (Argo) Rollout kinds, so determine which one is that
+func determineReplicaSetOwner(ownerName string, pod *corev1.Pod) (string, k8sconsts.WorkloadKind, error) {
+	// If we find a label associated with Argo rollouts, it is an Rollout kind
+	if _, ok := pod.Labels[argorolloutsv1alpha1.DefaultRolloutUniqueLabelKey]; ok {
```
since we only use the labels here, consider passing only the labels down the function chain

RonFed commented on opampserver/go.mod:
```diff
@@ -16,7 +16,7 @@ require (
 )
 
 require (
-	github.com/evanphx/json-patch v4.9.0+incompatible // indirect
```
a bit weird that it went away

natyharary replied:
Agreed. In general I don't get how `go mod tidy` decides to throw away versions but I can't modify it otherwise.
Should I manually re-add it? WDYT?

RonFed replied:
if its the result of go mod tidy, I guess its fine

natyharary commented on cli/cmd/describe.go:
(it is called `odigos-new` since I installed the new CLI temporarily as `odigos-new`)

natyharary commented on k8sutils/pkg/workload/ownerreference.go:
```diff
@@ -68,6 +71,16 @@ func GetWorkloadNameAndKind(ownerName, ownerKind string) (string, k8sconsts.Work
+// ReplicaSets can be created from either Deployment or (Argo) Rollout kinds, so determine which one is that
+func determineReplicaSetOwner(ownerName string, pod *corev1.Pod) (string, k8sconsts.WorkloadKind, error) {
+	// If we find a label associated with Argo rollouts, it is an Rollout kind
+	if _, ok := pod.Labels[argorolloutsv1alpha1.DefaultRolloutUniqueLabelKey]; ok {
```
It looks pretty ubiquitous throughout the argo rollouts code.
All argo rollouts (canary and bluegreen) use replicasets, and this label is always present in such replicaset

natyharary commented on instrumentor/controllers/sourceinstrumentation/instrumentationconfig_controller.go:
```diff
@@ -83,6 +83,12 @@ func getObjectByOwnerReference(ctx context.Context, k8sClient client.Client, own
 		return dc, err
 	}
 
+	if ownerRef.Kind == "Rollout" {
```
Good idea.

natyharary commented on instrumentor/controllers/sourceinstrumentation/manager.go:
```diff
@@ -32,6 +33,21 @@ func isDeploymentConfigAvailable(mgr ctrl.Manager) bool {
 
+// isRolloutAvailable checks if the Rollout resource is available in the cluster
+// using the RESTMapper to avoid permission errors on non-Rollout clusters
+func isRolloutAvailable(mgr ctrl.Manager) bool {
```
Done. I've moved the `GroupVersionKind` objects to k8sconsts like so:
```go
// GroupVersionKind constants for external workload types.
// These can be used to check resource availability in clusters.
var (
	DeploymentConfigGVK = schema.GroupVersionKind{
		Group:   "apps.openshift.io",
		Version: "v1",
		Kind:    string(WorkloadKindDeploymentConfig),
	}
	ArgoRolloutGVK = schema.GroupVersionKind{
		Group:   "argoproj.io",
		Version: "v1alpha1",
		Kind:    string(WorkloadKindArgoRollout),
	}
)
```
under `api/k8sconsts/podworkload.go`

natyharary committed:
fix: PR fixes - refactors and renames

blumamir commented on frontend/kube/client.go:
```diff
@@ -122,3 +128,35 @@ func IsDeploymentConfigAvailable() bool {
+		rolloutAvailable = false
+		rolloutAvailabilityChecked = true
+		return false
+	}
```
nit: there can be transient errors which are not related to permissions problems.

consider refining and adding log so we can know if there are any issues

natyharary replied:
Good one - added a log for tis

natyharary commented on cli/cmd/resources/ui.go:
ditto

natyharary commented on instrumentor/controllers/sourceinstrumentation/common.go:
```diff
@@ -44,6 +45,7 @@ func syncNamespaceWorkloads(
+		k8sconsts.WorkloadKindRollout,
 	} {
 		workloadObjects := workload.ClientListObjectFromWorkloadKind(kind)
 		err := k8sClient.List(ctx, workloadObjects, client.InNamespace(namespace))
```
Yes - the code below:
```go
if err != nil {
			// Ignore "no matches for kind" and "forbidden" errors for DeploymentConfig
			// This happens on non-OpenShift clusters where:
			// - The DeploymentConfig resource doesn't exist (NoMatchError)
			// - RBAC permissions aren't granted (Forbidden)
			if kind == k8sconsts.WorkloadKindDeploymentConfig && (meta.IsNoMatchError(err) || apierrors.IsForbidden(err)) {
				continue
			}
			// For other errors or other workload kinds, collect the error
			if !meta.IsNoMatchError(err) {
				errs = errors.Join(errs, err)
			}
			continue
		}
```

Catches any errors and logs them, in a nonblocking way.
Either way I added a case for the argo rollouts as well so there will not be any error logging if this kind is not in the cluster.

natyharary committed:
fix: RBAC, refactor IsArgoRolloutsAvailable

blumamir commented on k8sutils/pkg/utils/rollout.go:
```diff
@@ -176,6 +180,41 @@ func isDeploymentConfigRolloutDone(dc *openshiftappsv1.DeploymentConfig) bool {
+		// Waiting for rollout spec update to be observed
+		return false
+	}
+	if rollout.Spec.Replicas != nil && rollout.Status.UpdatedReplicas < *rollout.Spec.Replicas {
```
should we calculate replicas from spec?

blumamir commented on k8sutils/pkg/workload/workload.go:
```diff
@@ -24,6 +25,7 @@ var _ Workload = &StatefulSetWorkload{}
 var _ Workload = &CronJobWorkloadV1{}
 var _ Workload = &CronJobWorkloadBeta{}
 var _ Workload = &DeploymentConfigWorkload{}
+var _ Workload = &RolloutWorkload{}
```
ArgoRollout

RonFed commented on api/k8sconsts/podworkload.go:
```diff
@@ -1,5 +1,22 @@
 package k8sconsts
 
+import "k8s.io/apimachinery/pkg/runtime/schema"
```
kind of a nit, not sure if this this the right place for this? maybe should be in `k8sutils`? not sure about it

natyharary replied:
Perhaps in a future refactor when we move `WorkloadKind` entirely. I kind of just went with what worked with `WorkloadKindDeploymentConfig`

natyharary committed:
fix: refactors after code review

 IsArgoRollouts to be more generic
 remove mutex in FE + move const to k8sconsts
 isArgoRolloutRolloutDone refactors

blumamir commented on k8sutils/pkg/utils/rollout.go:
```diff
@@ -176,6 +180,27 @@ func isDeploymentConfigRolloutDone(dc *openshiftappsv1.DeploymentConfig) bool {
+	observedGen, err := strconv.ParseInt(rollout.Status.ObservedGeneration, 10, 64)
+	if err != nil || rollout.Generation > observedGen {
+		return false
+	}
```
check the observed generation before

blumamir commented on operator/bundle/manifests/odigos-operator.clusterserviceversion.yaml:
```diff
@@ -198,6 +198,7 @@ spec:
           resources:
           - pods
           verbs:
+          - delete
```
remove

blumamir commented on operator/bundle/manifests/odigos-operator.clusterserviceversion.yaml:
```diff
@@ -320,6 +321,18 @@ spec:
+          - rollouts
+          verbs:
+          - create
+          - delete
```
no create and delete

natyharary committed:
feat: CR RBAC fixes, e2e chainsaw tests for rollouts

RonFed reviewed (approved):
🔥 
I would consider moving the argo tests to the existing sources one

damemi commented on instrumentor/controllers/agentenabled/rollout/rollout.go:
```diff
@@ -238,6 +239,11 @@ func rolloutRestartWorkload(ctx context.Context, workload client.Object, c clien
+	case *argorolloutsv1alpha1.Rollout:
+		// Rollouts use a different field (spec.restartAt) for restarting, so we need to patch it differently
+		// https://github.com/argoproj/argo-rollouts/blob/cb1c33df7a2c2b1c2ed31b1ee0aa22621ef5577c/utils/replicaset/replicaset.go#L223-L232
+		rolloutPatch := []byte(fmt.Sprintf(`{"spec":{"restartAt":"%s"}}`, ts.Format(time.RFC3339)))
```
tbh, I wonder if we can switch to using a better annotation for all resource types here

afaik we just use `kubectl.kubernetes.io/restartedAt` because it's how kubectl restarts a deployment. but really, any annotation should work right? I think an odigos-specific annotation would be really helpful for showing what time *we* restarted the workloads wdyt about that @blumamir 

But still, in this case Argo rollouts don't just restart on an annotation update @natyharary ?

natyharary replied:
Argo rollouts work a bit differently here when it comes to `restartedAt` - instead of  `kubectl.kubernetes.io/restartedAt`, it restarts upon `spec.restartedAt`.

I might be missing something here regarding changing which annotation triggers which action, isn't it something rigid that cannot be changed? So Argo rollouts keep restarting upon `spec.restartedAt` and the rest of the workloads upon `kubectl.kubernetes.io/restartedAt`

> But still, in this case Argo rollouts don't just restart on an annotation update @natyharary ?

@damemi  When I checked, Argo rollouts didn't restart upon `kubectl.kubernetes.io/restartedAt`

damemi commented on k8sutils/pkg/client/client.go:
```diff
@@ -97,3 +99,10 @@ func GetK8sClientset() (*kubernetes.Clientset, error) {
+
+// IsResourceAvailable checks if the external resource is available in the cluster
+// using the RESTMapper to avoid permission errors on clusters that don't have the resource installed.
+func IsResourceAvailable(mapper meta.RESTMapper, gvk schema.GroupVersionKind) bool {
```
awesome
